### PR TITLE
[pytx] Implement a new cleaner PDQ index solution

### PR DESCRIPTION
**Pull request by haianhng31**
Summary
---------
Resolve issue #1613 . 
This PR introduces a new PDQIndex2 class for managing and querying a PDQ hash-based index using FAISS. Key changes include:
* **_PDQHashIndex**: A wrapper around the FAISS index to handle adding hashes and performing searches using the FAISS library.
* **PDQIndex2**: A class for managing a PDQ index, which includes:
  * The ability to add PDQ hashes and associate them with entries. 
  * A query method for searching the index for matching entries based on a query hash. 
  * A serialize and deserialize method for persisting and loading the index from binary streams using pickle. 

Test Plan
---------
I have included several test cases for this, currently all in one file.

For the bug in issue #1318, I have created 2 test cases relating to it:
* [Test how the index handles when it only has one entry](https://github.com/facebook/ThreatExchange/pull/1698/files#diff-52467a9eead5544de4d6082c507dd5c96bebe25c34a1a5fdfebbdb9f2c223846R182-R195). -> Return correctly 1 hash
* [Test exact matches in sample set](https://github.com/facebook/ThreatExchange/pull/1698/files#diff-52467a9eead5544de4d6082c507dd5c96bebe25c34a1a5fdfebbdb9f2c223846R109-R120) -> Doesn't return all hashes

Also ran `test_pdq_index` with 1,000 banked & 10,000 queries in the unit test. The test passed but it took 4 minutes to run.
And if i tried with 10,000 banked & 100,000 queries, the test took forever.

### CODE REVIEW

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        Serialize the PDQ index to a binary stream.
+        """
+        fout.write(pickle.dumps(self))
+
+    @classmethod
+    def deserialize(cls, fin: t.BinaryIO) -> "PDQIndex2[IndexT]":
+        """
+        Deserialize a PDQ index from a binary stream.
+        """
+        return pickle.loads(fin.read())
```
Comment (haianhng31) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:108]: Q: Since these methods already exist in SignalTypeIndex, do we still need to add them here?

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+    def _convert_pdq_strings_to_ndarray(
+        self, pdq_strings: t.Sequence[str]
+    ) -> np.ndarray:
+        """
+        Convert multiple PDQ hash strings to a numpy array.
+        Args:
+            pdq_strings (Sequence[str]): A sequence of 64-character hexadecimal PDQ hash strings
+        Returns:
+            np.ndarray: A 2D array of shape (n_queries, 256) where each row is the full PDQ hash as a bit array
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:166]: I don't think this return type is accurate, since I think it needs to be bytes for binary index

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                hash_int = int(pdq_str, 16)
+                # Convert to binary string, padding to ensure 256 bits
+                binary_str = format(hash_int, "0256b")
+                # Convert to numpy array
+                hash_array = np.array(
+                    [int(bit) for bit in binary_str], dtype=np.float32
+                )
+                hash_arrays.append(hash_array)
+            except (ValueError, TypeError) as e:
+                raise ValueError(f"Invalid PDQ hash string: {pdq_str}") from e
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:182]: unsure: What's the difference between doing this and letting the exception leak out except coercing TypeError?

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+        # Convert list of arrays to a single 2D array
+        return np.array(hash_arrays, dtype=np.float32)
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:185]: This is the wrong type, should be `uint8`

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    T as IndexT,
+)
+from threatexchange.signal_type.pdq.pdq_faiss_matcher import (
+    PDQMultiHashIndex,
+    PDQFlatHashIndex,
+    PDQHashIndex,
+)
+
+DEFAULT_MATCH_DIST = 31
+DIMENSIONALITY = 256
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:27]: blocking: These look like they are copied from somewhere else in the code - we should only have one set of these constants.

If you are running into a circular import with PDQSignal, then move the constants into their own file

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+class PDQIndex2(SignalTypeIndex[IndexT]):
+    """
+    Wrapper around the pdq faiss index lib using PDQMultiHashIndex
+    """
+
+    def __init__(
+        self,
+        threshold: int = DEFAULT_MATCH_DIST,
+        index: t.Optional[faiss.Index] = None,
+        entries: t.Iterable[t.Tuple[str, IndexT]] = (),
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:41]: I think we can probably leave this off and rely on `build`

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+class PDQIndex2(SignalTypeIndex[IndexT]):
+    """
+    Wrapper around the pdq faiss index lib using PDQMultiHashIndex
+    """
+
+    def __init__(
+        self,
+        threshold: int = DEFAULT_MATCH_DIST,
+        index: t.Optional[faiss.Index] = None,
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:40]: nit: swap these and turn threshold into a kw arg

```suggestion
        index: t.Optional[faiss.Index] = None,
        *
        threshold: int = DEFAULT_MATCH_DIST,

```

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+DEFAULT_MATCH_DIST = 31
+DIMENSIONALITY = 256
+
+PDQIndexMatch = IndexMatchUntyped[SignalSimilarityInfoWithIntDistance, IndexT]
+
+
+class PDQIndex2(SignalTypeIndex[IndexT]):
+    """
+    Wrapper around the pdq faiss index lib using PDQMultiHashIndex
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:34]: bl;ocking: Comment out of date?

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        self,
+        threshold: int = DEFAULT_MATCH_DIST,
+        index: t.Optional[faiss.Index] = None,
+        entries: t.Iterable[t.Tuple[str, IndexT]] = (),
+    ) -> None:
+        super().__init__()
+        self.threshold = threshold
+
+        if index is None:
+            index = faiss.IndexFlatL2(DIMENSIONALITY)
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:47]: 👍

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        self.index = _PDQHashIndex(index)
+
+        # Matches hash to Faiss index
+        self._deduper: t.Dict[str, faiss.IndexFlatL2] = {}
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:51]: blocking: This is the wrong type on the value. I believe it should be `int`

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    def search(
+        self, queries: t.Sequence[str], threshold: int = DEFAULT_MATCH_DIST
+    ) -> t.List[t.List[t.Any]]:
+        """
+        Search the FAISS index for matches to the given PDQ queries.
+        Args:
+            queries (Sequence[str]): The PDQ signal strings to search for.
+            threshold (int): The maximum distance threshold for matches.
+        Returns:
+            2D list of tuples that store (matches, distances) for each query
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:137]: blocking: This doesn't match the type annotation for the method. Suggest not mentioning the type and focus on the intent, since the code types can document the types

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+        self.faiss_index.add(vectors)
+
+    def search(
+        self, queries: t.Sequence[str], threshold: int = DEFAULT_MATCH_DIST
+    ) -> t.List[t.List[t.Any]]:
+        """
+        Search the FAISS index for matches to the given PDQ queries.
+        Args:
+            queries (Sequence[str]): The PDQ signal strings to search for.
+            threshold (int): The maximum distance threshold for matches.
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:135]: nit: Recommend removing types from these, since it only creates the possibility they are out of sync with the actual types

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+        Add PDQ hashes to the FAISS index.
+        Args:
+            pdq_strings (Sequence[str]): PDQ hash strings to add
+        """
+        vectors = self._convert_pdq_strings_to_ndarray(pdq_strings)
+        self.faiss_index.add(vectors)
+
+    def search(
+        self, queries: t.Sequence[str], threshold: int = DEFAULT_MATCH_DIST
+    ) -> t.List[t.List[t.Any]]:
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:130]: You can narrow this typing down more I think, tuple(int, int)

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        query_array: np.ndarray = self._convert_pdq_strings_to_ndarray(queries)
+        limits, distances, indices = self.faiss_index.range_search(
+            query_array, threshold + 1
+        )
+
+        results: t.List[t.List[t.Any]] = []
+        for i in range(len(queries)):
+            matches = [idx.item() for idx in indices[limits[i] : limits[i + 1]]]
+            dists = [dist for dist in distances[limits[i] : limits[i + 1]]]
+            results.append(list(zip(matches, dists)))
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:148]: Do you need to use `tuple` instead of `list` here?

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+        return results
+
+    def __getstate__(self):
+        data = faiss.serialize_index(self.faiss_index)
+        return data
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:153]: nit: inline
```suggestion
        return faiss.serialize_index(self.faiss_index)
```

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+    def __setstate__(self, data):
+        self.faiss_index = faiss.deserialize_index(data)
+
+    def _convert_pdq_strings_to_ndarray(
+        self, pdq_strings: t.Sequence[str]
+    ) -> np.ndarray:
+        """
+        Convert multiple PDQ hash strings to a numpy array.
+        Args:
+            pdq_strings (Sequence[str]): A sequence of 64-character hexadecimal PDQ hash strings
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:164]: nit: on leaving off types

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+    return index, pdq_hashes
+
+
+def test_init(empty_index) -> None:
+    assert empty_index.threshold == DEFAULT_MATCH_DIST
+    assert isinstance(empty_index.index, _PDQHashIndex)
+    assert isinstance(empty_index.index.faiss_index, faiss.IndexFlatL2)
+    assert empty_index.index.faiss_index.d == DIMENSIONALITY
+    assert empty_index._deduper == dict()
+    assert empty_index._idx_to_entries == []
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:62]: Suggest leaving these off - it doesn't really test anything

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+        entries=[
+            ("f8f8f0cee0f4a84f06370a22038f63f0b36e2ed596621e1d33e6b39c4e9c9b22", 0)
+        ]
+    )
+    return index, pdq_hashes
+
+
+def test_init(empty_index) -> None:
+    assert empty_index.threshold == DEFAULT_MATCH_DIST
+    assert isinstance(empty_index.index, _PDQHashIndex)
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:58]: suggest leaving this off - doesn't test anything

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+    assert isinstance(empty_index.index.faiss_index, faiss.IndexFlatL2)
+    assert empty_index.index.faiss_index.d == DIMENSIONALITY
+    assert empty_index._deduper == dict()
+    assert empty_index._idx_to_entries == []
+
+
+def test_serialize_deserialize(empty_index) -> None:
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:65]: blocking: This test doesn't really test much unless the index has some contents.

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+    assert isinstance(deserialized_index, PDQIndex2)
+    assert deserialized_index.threshold == empty_index.threshold
+    assert isinstance(deserialized_index.index, _PDQHashIndex)
+    assert isinstance(deserialized_index.index.faiss_index, faiss.IndexFlatL2)
+    assert deserialized_index.index.faiss_index.d == DIMENSIONALITY
+    assert deserialized_index._deduper == empty_index._deduper
+    assert deserialized_index._idx_to_entries == empty_index._idx_to_entries
+
+
+def test_serialize_deserialize_with_custom_index_threshold(
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:80]: blocking: This test also doesn't test much, we can probably leave it off

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+    assert len(results) == 0
+
+
+def test_sample_set_near_match(sample_index):
+    """Test near matches in sample set."""
+    index, pdq_hashes = sample_index
+
+    # Create a near-match by flipping a few bits
+    base_hash = pdq_hashes[0]
+    near_hash = hex(int(base_hash, 16) ^ 0xF)[2:].zfill(64)  # Flip 4 bits
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:136]: blocking: Did you forget to query near_hash?

ignorable: In another similar project, I wrote a helper on the signal type itself to provide a random hash that was X distance away from a given hash by flipping random bits.

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+
+    results = index.query(near_hash)
+    assert len(results) > 0  # Should find near matches
+    assert results[0].similarity_info.distance > 0
+
+
+def test_sample_set_threshold(sample_index):
+    """Test distance threshold behavior."""
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:144]: nit: Try to avoid docs that are tautological with the function name

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+        "f" * 64,
+        "0" * 64,
+        "a" * 64,
+    ]
+    index = PDQIndex2[str](
+        entries=[
+            ("f8f8f0cee0f4a84f06370a22038f63f0b36e2ed596621e1d33e6b39c4e9c9b22", 0)
+        ]
+    )
+    return index, pdq_hashes
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:53]: blocking: Why do you return hashes which aren't in the index?

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+    _PDQHashIndex,
+    DIMENSIONALITY,
+    DEFAULT_MATCH_DIST,
+)
+
+
+@pytest.fixture
+def empty_index():
+    """Fixture for an empty index."""
+    return PDQIndex2[str]()
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:15]: blocking: Why subscript this?
```suggestion
    return PDQIndex2()
```

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+def sample_index():
+    """Fixture for an index with a small sample set."""
+    pdq_hashes = [
+        "f8f8f0cee0f4a84f06370a22038f63f0b36e2ed596621e1d33e6b39c4e9c9b22",
+        "f" * 64,
+        "0" * 64,
+        "a" * 64,
+    ]
+    index = PDQIndex2[str](entries=[(h, pdq_hashes.index(h)) for h in pdq_hashes])
+    return index, pdq_hashes
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:36]: blocking: I'm not sure about this pattern of returning an index and the sample hashes. It doesn't seem like you are saving that much time vs something like this:

```
SAMPLE_HASH = "f8f8f0cee0f4a84f06370a22038f63f0b36e2ed596621e1d33e6b39c4e9c9b22"

SAMPLE_HASHES = [
  SAMPLE_HASH,
  "f" * 64,
  "0" * 64,
  "a" * 64,
]

@dataclass
class IndexRecord:
  original_hash: string
  idx: int 

  def get_expected_distance(matched_hash) -> int:
    return pdq_utils.simple_distance_binary(self.original_hash, matched_hash)

def make_index(hashes):
  return PDQIndex2.build((h, StoredRecord(h, i) for i, h in enumerate(hashes)

def test_something():
  index = make_index(SAMPLE_HASHES)

  query_hash = "..."

  for d, record in index.search([query_hash ]):
    assert d == record.expected_distance(query_hash)
  
```

And you also aren't consistent on the test fixtures about what will be in the return. Fixtures are nice, but they aren't mandatory

**Review by Dcallies — CHANGES_REQUESTED** [review]
Looking good so far!

Toplevel comments:
1. I am not convinced yet that this indexing solution is correct yet. To help convince me, you can write a unittest that generates two large sets of random hashes (PDQSignal.get_random_signal()) b and q (e.g. 1,000 and 10,000), uses the brute force method to check exactly how many of hashes in b should match the hashes in q, then loads the b hashes into your indexing solution, and then queries all of the q hashes and makes sure you exactly get the expected results. 
1. Many of your existing tests are tautological or should check to make sure the matching hashes are actually the ones that should be matching.  You can likely replace many of your existing tests with variations on 2 above - I put some ideas for you in the IndexRecord class, but there are others that might work.
1. I recommend adding a method that can return a random hash exactly X bits different from a given hash, and then using that to help in your tests.

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py (hunk 23 above)
Comment (haianhng31) [reply python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:53]: so that I can use the other hashes to test for unmatch when needed. 
For example, in the test: `test_one_entry_sample_index`

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py (hunk 21 above)
Comment (haianhng31) [reply python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:136]: I query the near_hash on the [line](https://github.com/facebook/ThreatExchange/pull/1698/files#diff-52467a9eead5544de4d6082c507dd5c96bebe25c34a1a5fdfebbdb9f2c223846R138) below

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+from threatexchange.signal_type.index import (
+    IndexMatchUntyped,
+    SignalSimilarityInfoWithIntDistance,
+    SignalTypeIndex,
+    T as IndexT,
+)
+from threatexchange.signal_type.pdq.pdq_faiss_matcher import (
+    PDQMultiHashIndex,
+    PDQFlatHashIndex,
+    PDQHashIndex,
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:23]: We don't want to import these because they are the code you are going to be replacing!

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+        bits[pos] = "0" if bits[pos] == "1" else "1"  # Flip selected bit positions
+    modified_hash = hex(int("".join(bits), 2))[2:].zfill(64)  # Convert back to hex
+
+    return modified_hash
+
+
+def test_pdq_index():
+    # Make sure base_hashes and query_hashes have at least 100 similar hashes
+    base_hashes = SAMPLE_HASHES + [PdqSignal.get_random_signal() for _ in range(1000)]
+    query_hashes = SAMPLE_HASHES + [PdqSignal.get_random_signal() for _ in range(10000)]
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:47]: Why add SAMPLE_HASHES here? If you are trying to guarantee matches, you could generate less query hashes and copy the banked hashes in

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 28]
```diff
@@ -0,0 +1,138 @@
...
+
+def test_pdq_index_with_exact_distance():
+    thresholds: t.List[int] = [10, 31, 50]
+    indexes: t.List[PDQIndex2] = []
+    for thres in thresholds:
+        index = PDQIndex2(
+            entries=[(h, SAMPLE_HASHES.index(h)) for h in SAMPLE_HASHES],
+            threshold=thres,
+        )
+        indexes.append(index)
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:78]: ignorable: Could do this as a comprehension

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 29]
```diff
@@ -0,0 +1,138 @@
...
+    results = index.query(SAMPLE_HASHES[0])
+
+    # Should find all entries associated with the hash
+    assert len(results) == 4
+    for result in results:
+        assert result.similarity_info.distance == 0
+
+
+def test_one_entry_sample_index():
+    """Test how the index handles when it only has one entry."""
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:126]: nit: Add a reference to the issue

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+    base: t.List[str], query: str, threshold: int = 32
+) -> t.Set[int]:
+    matches = set()
+    query_arr = convert_pdq_strings_to_ndarray([query])[0]
+
+    for i, base_hash in enumerate(base):
+        base_arr = convert_pdq_strings_to_ndarray([base_hash])[0]
+        distance = np.count_nonzero(query_arr != base_arr)
+        if distance <= threshold:
+            matches.add(i)
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py:22]: blocking: Why create a new version of this instead of using existing methods like `simple_distance` or PDQSignal.compare_hash`?

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py [hunk 31]
```diff
@@ -0,0 +1,132 @@
...
+    convert_pdq_strings_to_ndarray,
+)
+
+PDQIndexMatch = IndexMatchUntyped[SignalSimilarityInfoWithIntDistance, IndexT]
+
+
+class PDQIndex2(SignalTypeIndex[IndexT]):
+    """
+    Indexing and querying PDQ signals using Faiss for approximate nearest neighbor search.
+    """
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_index2.py:31]: Nit:

Add a note that the `2` is because it's a redo of the existing PDQ index designed to be simpler and fix hard-to-squash bugs in the existing implementation. The intention is that this class will replace the original index in pytx 2.0

File: python-threatexchange/threatexchange/signal_type/pdq/pdq_utils.py [hunk 32]
```diff
@@ -49,3 +52,19 @@ def pdq_match(pdq_hex_a: str, pdq_hex_b: str, threshold: int) -> bool:
...
+
+def convert_pdq_strings_to_ndarray(pdq_strings: t.Sequence[str]) -> np.ndarray:
+    """
+    Convert multiple PDQ hash strings to a numpy array.
+    """
+    if not all(len(pdq_str) == PDQ_HEX_STR_LEN for pdq_str in pdq_strings):
+        raise ValueError("All PDQ hash strings must be 64 hex characters long")
+
+    binary_arrays = []
+    for pdq_str in pdq_strings:
```
Comment (Dcallies) [inline python-threatexchange/threatexchange/signal_type/pdq/pdq_utils.py:65]: ignorable: If you move the ValueError check inside the loop you can change the type of pdq_strings to be `t.Iteratble` and accept generators.

**Review by Dcallies — CHANGES_REQUESTED** [review]
So close!

1. The way you are seeding your hashes is causing duplicates in your query set - see inline blocking comment
2. As part of your test plan (doesn't have to be part of the test itself), please try generate a very large number of hashes and queries - e.g. 10,000 banked and 100,000 queries - something that takes 5-10 seconds to check. We want to be certain that the index is correct!
3. Please add a test that serializes and deserializes a non-empty index - you can do something cheeky like:
```
original = ... # do stuff to prepare index
deserialized = PDQIndex2,deserialize(index.serialize(...))
for index in (original, deserialized):
   # The original test goes here, should get same results on original or deserialized
```

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py [hunk 33]
```diff
@@ -0,0 +1,144 @@
+import typing as t
+import numpy as np
+import random
+
+from threatexchange.signal_type.pdq.pdq_index2 import PDQIndex2
+from threatexchange.signal_type.pdq.signal import PdqSignal
+from threatexchange.signal_type.pdq.pdq_utils import simple_distance
+
+
+def _generate_sample_hashes(size: int, seed: int = 42):
```
[the remark this answers is not in the document]
Comment (haianhng31) [reply python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py]: I'm sorry what do you mean with the image:P ?

**Comment by haianhng31** [comment]
@Dcallies 
I have tried with 1,000 banked & 10,000 queries in the unit test. The test passed but it took 4 minutes to run. 
And if i tried with 10,000 banked & 100,000 queries, the test took forever.

File: python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py (hunk 33 above)
[the remark this answers is not in the document]
Comment (Dcallies) [reply python-threatexchange/threatexchange/signal_type/tests/test_pdq_index2.py]: 42 is the "Answer to the Ultimate Question of Life, The Universe, and Everything" in the book series "The Hitchhiker's Guide to the Galaxy", of which the image is on the cover. I thought you were a fan :P 

Read more: https://en.wikipedia.org/wiki/Phrases_from_The_Hitchhiker%27s_Guide_to_the_Galaxy
